foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/meter/BucketsDistribution.java

```diff
@@ -0,0 +1,162 @@
+ * limitations under the License.
+ */
+package org.apache.servicecomb.foundation.metrics.meter;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.LongAdder;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.netflix.config.DynamicPropertyFactory;
+import com.netflix.spectator.api.Id;
+import com.netflix.spectator.api.Measurement;
+
+public class BucketsDistribution extends AbstractPeriodMeter {
+  private static final Logger LOGGER = LoggerFactory.getLogger(BucketsDistribution.class);
+
+  public static final String METRICS_LATENCY = DynamicPropertyFactory.getInstance()
+      .getStringProperty("servicecomb.metrics.invocation.latency.buckets", "0")
+      .get();
+
+  private List<LatencyInterval> latencyIntervals;
+
+  public BucketsDistribution(Id id) {
+    this.id = id;
+    this.latencyIntervals = parseLatencyIntervalListFromProperty(METRICS_LATENCY, id);
+  }
+
+  public void record(long nanoAmount) {
+    if (nanoAmount > 0) {
+      for (LatencyInterval latencyInterval : latencyIntervals) {
+        if (latencyInterval.match(nanoAmount)) {
+          latencyInterval.increment();
+          return;
+        }
+      }
+    }
+  }
+
+  @Override
+  public void calcMeasurements(long msNow, long secondInterval) {
+    List<Measurement> measurements = new ArrayList<>();
+    calcMeasurements(measurements, msNow, secondInterval);
+    allMeasurements = measurements;
+  }
+
+  @Override
+  public void calcMeasurements(List<Measurement> measurements, long msNow, long secondInterval) {
+    latencyIntervals.forEach(latencyInterval -> {
+      measurements.add(newMeasurement(latencyInterval.getId(),
+          msNow, latencyInterval.getAndUpdateLastTimes()));
+    });
+  }
+
+  private Measurement newMeasurement(Id id, long msNow, Number n) {
+    return new Measurement(id, msNow, n.doubleValue());
+  }
+
+  public static List<LatencyInterval> parseLatencyIntervalListFromProperty(String property, Id rootId) {
+    List<LatencyInterval> results = new ArrayList<>();
+    try {
+      int[] arrays = Arrays.stream(property.trim().split(","))
+          .mapToInt(Integer::parseInt)
+          .distinct()
+          .sorted()
+          .filter(num -> num > 0)
+          .toArray();
+      if (arrays == null || arrays.length == 0) {
+        return results;
+      }
+      results.add(new LatencyInterval(0L, arrays[0], rootId));
+      for (int i = 1; i < arrays.length; i++) {
+        results.add(new LatencyInterval(arrays[i - 1], arrays[i], rootId));
+      }
+      results.add(new LatencyInterval(arrays[arrays.length - 1], Long.MAX_VALUE, rootId));
+    } catch (Throwable e) {
+      LOGGER.error("Failed to parse property servicecomb.metrics.invocation.latency.buckets", e);
+    }
+    return results;
+  }
+
+  public static class LatencyInterval {
+
+    public static final int leastLength = 7;
```

Comment:
only for publish, not relate to calc......

Reply:
I have moved the **leastLength**  to the class **DefaultLogPublisher**.
as I thought that this number is relate to the log and has no relation with tag

```java
public class DefaultLogPublisher implements MetricsInitializer {
  private static final int LEAST_LATENCY_SCOPE_STR_LENGTH = 7;
```